lib/pbench/cli/server/shell.py

```diff
@@ -40,42 +42,135 @@ def find_the_unicorn(logger: Logger):
+        self.logger = logger
+        self.indexer: Optional[subprocess.Popen] = None
+        self.reindexer: Optional[subprocess.Popen] = None
+        self.shutoff: Event = Event()
+        self.reaper: Optional[Thread] = None
 
-    Return 0 on success, 1 on failure.
-    """
-    ret_val = 0
-    crontab_f = Path(crontab_dir) / "crontab"
-    if not crontab_f.exists():
-        os.environ["PATH"] = ":".join([str(bin_dir), os.environ["PATH"]])
-        # Create the crontab file from the server configuration.
-        cp = subprocess.run(["pbench-create-crontab", crontab_dir], cwd=cwd)
-        if cp.returncode != 0:
-            logger.error(
-                "Failed to create crontab file from configuration: {}", cp.returncode
-            )
-            ret_val = 1
-        else:
-            # Install the created crontab file.
-            cp = subprocess.run(["crontab", f"{crontab_dir}/crontab"], cwd=cwd)
-            if cp.returncode != 0:
-                logger.error("Failed to install crontab file: {}", cp.returncode)
-                ret_val = 1
-    if ret_val != 0:
-        crontab_f.unlink(missing_ok=True)
-    return ret_val
-
-
-def run_gunicorn(server_config: PbenchServerConfig, logger: Logger) -> int:
+    def start_indexer(self):
+        """Isolate starting the primary indexer as a subprocess.
+
+        NOTE: we specify the 'python3' command explicitly rather than relying
+        on the shebang to avoid the intermediate `env python` command which
+        complicates reaping.
+        """
+        self.indexer = subprocess.Popen(
+            ["python3", self.bin_dir / "pbench-index"], cwd=self.cwd
+        )
+        self.logger.info("Index process running in pid {}", self.indexer.pid)
+
+    def start_reindexer(self):
+        """Isolate starting the secondary reindexer as a subprocess.
+
+        TODO: I'm not sure we really want to retain the old "reindex". For one
+        thing, re-indexing with existing index documents may be problematic,
+        especially if the new indexer doesn't overwrite every document. A
+        better mechanism would be to delete all the existing documents using
+        the index-map and then we can simply READY the INDEX operation.
+
+        NOTE: we specify the 'python3' command explicitly rather than relying
+        on the shebang to avoid the intermediate `env python` command which
+        complicates reaping.
+        """
+        self.reindexer = subprocess.Popen(
+            ["python3", self.bin_dir / "pbench-index", "--re-index"], cwd=self.cwd
+        )
+        self.logger.info("Re-index process running in pid {}", self.reindexer.pid)
+
+    def watch_indexers(self):
+        """A thread to monitor our two indexer subprocesses.
+
+        It'll periodically check `wait` for terminated processes in the process
+        group. If the pid matches the indexer or reindexer, restart them. If a
+        gunicorn work fails (or any other subprocess that might occur) just log
```

Comment:
FWIW, we could use systemd timers to invoke the indexer instead of handling it ourselves.

Reply:
Ah; so I should just throw this entire PR out the window. Well, that's one thought; and it would presumably deal with the whole issue of handling dead processes.

The one real advantage here is that we're tying them to the server lifecycle, which reflects the current `crontab` integration. But I suppose we could also run `systemctl` from `shell.py` to enable and disable our timed execution, so that may not be a big deal.

Reply:
I think you have systemd handle all the dependencies so that when the pbench-server is "stopped" the timers are "stopped", and when the pbench-server is "started" the timers are "started".

Reply:
> I think you have systemd handle all the dependencies so that when the pbench-server is "stopped" the timers are "stopped", and when the pbench-server is "started" the timers are "started".

Well that's worth some investigation ...

Reply:
I am investigating this right now as I work on a proposal for unification for the "to-be-created" `b0.72` branch to incorporate the old code in a `server/legacy` sub-directory.